docs/src/crypto-primitives/tsmt.md

```diff
@@ -0,0 +1,123 @@
+
+The first documented implementation of a compact SMT was described [here](https://github.com/proofchains/python-proofmarshal/blob/master/proofmarshal/merbinnertree.py). This implementation relies on augmenting nodes with
+a field called `prefix` which is the longest common prefix of all childs of a given node. A disadvantage of this implementation, as well as all known implementations of compact SMTs, is the reliance of their operations on bit-wise manipulations. This makes them very expensive to implement inside Miden VM.
+
+Tiered sparse Merkle trees (TSMT) are a result of the idea that one can trade hashing for more efficient bit-wise operations. More precisely, in TSMT, compaction happens only at certain pre-specified tiers and this leads to simpler bit-wise manipulations. In the case of Miden Vm, compaction happens at depths `16`, `32`, `48` and `64`. To illustrate this, consider the following example where the keys have length 4 i.e. the SMT has depth 4.
+
+<p align="center">
+    <img src="./diagram/../../diagrams/crypto-primitives/tsmt/tsmt_non_compact.svg">
+</p>
+
+If we choose the tiers to be at depths 2 and 4 then the compact version of the above tree will look like
+
+<p align="center">
+    <img src="./diagram/../../diagrams/crypto-primitives/tsmt/tsmt_compact_2.svg">
+</p>
+
+If, however, we choose compaction at all possible depths we get a compact SMT that is similar to the [Jellyfish SMT](https://developers.diem.com/papers/jellyfish-merkle-tree/2021-01-14.pdf)
+
+<p align="center">
+    <img src="./diagram/../../diagrams/crypto-primitives/tsmt/tsmt_compact_1.svg">
+</p>
+
+We can see that a TSMT is a parametrized SMT, where the parameter is the number of tiers, that interpolates between a plain SMT with no compaction and a  (fully) compact SMT with compaction allowed at evey depth.
+
+In Miden VM, the TSMT has a key space of size approximately `256` and in order to deal with keys sharing a common prefix of size greater than `64` we introduce a special type of leaf nodes at depth `64`. These leaf nodes are a sorted list of pairs `(key, value)` and sorting is done using `key`. Notice that all pairs in the sorted list have the same first part of the key, namely they share a common `64` bit prefix, and that all pairs with the same first part of the key belong in the same sorted list. The decision to go with a sorted list of pairs instead of a standard sub-tree node with multiple childs at depth `80` sharing the same `64` bit prefix is partially motivated by the fact that, for most scenarios related to Miden VM, the probability that two distinct keys will end up sharing a common prefix of length greater than or equal to `64` is very low, and even feasable grinding attacks can make the sorted list only slightly longer than a dozen of elements with very little clear benefits. Thus putting all pairs with the same first part of the key in a sorted list is a good tradeoff for dealing with prefixes of length greater than `64`.
```

Comment:
- In plural sounds better to me but I am not 100% sure.
- Thanks!

Reply:
Me neither; researching it a bit, it should be singular because it might be an uncountable noun as your reference on the idea of a node. But I am not even sure if this is right or important. It just stood out when reading it.

These language nerds, so to speak, might add clarity. 

https://english.stackexchange.com/questions/29843/types-of-followed-by-singular-or-plural

Reply:
Thanks for the link but I think it talks about `types of` and not `type of`. 
However, I think that both can be used only that if we use `type of leaf node` then the next sentence will have to be changed to something like `This type of nodes is a sorted list` which sounds a bit heavy for my taste.
In any case, we can leave it as is, if you agree, as it doesn't affect the meaning very significantly.

Reply:
You are absolutely right; let's keep it as is.